lib/puppet/agent.rb

```diff
@@ -60,6 +60,12 @@ def run(client_options = {})
               end
             end
           rescue Puppet::LockError
+            if client_options[:waitforlock]
+              while File.file?(lockfile_path)
+                sleep 0.5
+              end
+              retry
+            end
```

Comment:
Probably don't want to wait indefinitely. Maybe add a `maxwaitforlock` puppet setting, so that it can be used in other places where we need to acquire lock files, such as the ssl lockfile in `puppet ssl bootstrap`.